gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/TokenMAC.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.knox.gateway.services.token.impl;
+
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+
+import javax.crypto.Mac;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
+
+import org.apache.knox.gateway.config.GatewayConfig;
+import org.apache.knox.gateway.services.ServiceLifecycleException;
+import org.apache.knox.gateway.services.security.AliasService;
+import org.apache.knox.gateway.services.security.AliasServiceException;
+
+public class TokenMAC {
+  static final String KNOX_TOKEN_HASH_KEY_ALIAS_NAME = "knox.token.hash.key";
+
+  private final Mac tokedMac;
+
+  TokenMAC(GatewayConfig config, AliasService aliasService) throws ServiceLifecycleException {
+    if (aliasService == null) {
+      throw new ServiceLifecycleException("Missing alias service while configuring Knox Token MAC.");
+    }
+
+    try {
+      final String algorithm = config.getKnoxTokenHashAlgorithm();
+      final char[] knoxTokenHashKey = aliasService.getPasswordFromAliasForGateway(KNOX_TOKEN_HASH_KEY_ALIAS_NAME);
+      if (knoxTokenHashKey != null) {
+        final SecretKey key = new SecretKeySpec(new String(knoxTokenHashKey).getBytes(StandardCharsets.UTF_8), algorithm);
+        tokedMac = Mac.getInstance(algorithm);
+        tokedMac.init(key);
+      } else {
+        throw new ServiceLifecycleException("Missing " + KNOX_TOKEN_HASH_KEY_ALIAS_NAME + " alias from Gateway's credential store");
```

Comment:
Why not fallback on master secret?

Reply:
The master secret cannot be used as key length matters.